File: parser/fight_extractor.py
# parser/fight_extractor.py
from __future__ import annotations

import re
import sys
from collections import deque
from typing import Optional

from .types import Event, Fight
from .bosses import build_phases_for_fight
from .boss_match import match_boss_name, boss_involved

DEATH_RE = re.compile(r"est\s+mort", re.IGNORECASE)        # "X est mort(e)."
KILL_RE = re.compile(r"\ba\s+tu[ée]\b", re.IGNORECASE)     # "A a tué B"

# Uniquement des mots typiques de lignes de dégâts/soins (réduit fortement les faux positifs)
COMBAT_WORD_RE = re.compile(
    r"(inflige|dég[âa]ts?|damage|touche|frappe|attaque|crit|critique|soigne|heal)",
    re.IGNORECASE,
)
HAS_NUMBER_RE = re.compile(r"\b\d+\b")


def extract_kills(events: list[Event]) -> list[Fight]:
    """
    Kill "commun" :
      - code 11 ou texte "a tué" => dst est mort
      - code 12 + "est mort" => src est mort
      - on ne compte que si l'entité morte est un boss

    IMPORTANT :
    - Combat Begin/End peut être local (mort/phase/aggro) => ne doit PAS reset un boss fight
      si Begin suit un End de près (stitch).

    FIX "ligne parasite" (ex: Général Silgen) :
    - un code 12 "Boss est mort" peut apparaître trop tôt.
    - on ne finalise PAS immédiatement un kill sur code 12.
      => pending + confirmation plus tard si le boss ne réapparaît plus en combat
         ou si un code 11 arrive (prioritaire).

    REGLE "NE PAS COMPTER" (sans remonter le start) :
    - Au moment d'un boss kill (code 11 / "a tué"), on regarde s'il y a eu des
      interactions *réellement combat* avec CE boss dans les 3 secondes AVANT le Combat Begin.
    - Si oui => on NE COMPTE PAS ce kill, et on jette le pull courant.
    - On log explicitement la raison sur stderr.
    """

    fights: list[Fight] = []
    current_start: Optional[Event] = None
    current_events: list[Event] = []
    kill_idx = 0

    last_combat_end: Optional[Event] = None
    last_boss_seen_sec: Optional[int] = None

    # Stitch/rejoin
    REJOIN_GAP_SEC = 90
    BOSS_RECENT_SEC = 180

    # Anti-parasite code 12
    CONFIRM_DEATH_AFTER_SEC = 6
    pending_dead_boss: Optional[str] = None
    pending_dead_event: Optional[Event] = None
    pending_invalidated: bool = False

    # Lookback global (avant Begin)
    LOOKBACK_BEFORE_BEGIN_SEC = 3
    GLOBAL_BUFFER_SEC = 12  # >= LOOKBACK_BEFORE_BEGIN_SEC
    MIN_COMBATLIKE_INTERACTIONS = 2  # comme ton exemple (2 interactions dans la fenêtre)

    recent_events: deque[Event] = deque()

    # --- SPECIAL: Council of Fate tracking ---
    council_dead: set[str] = set()

    COUNCIL_KEY = "Le Concile du Destin"
    COUNCIL_MEMBERS = {
        "countessa danazhal",
        "marquise boldoch",
        "comte pluezhal",
    }

    def _is_council_member(name: str) -> bool:
        return (name or "").strip().casefold() in COUNCIL_MEMBERS
    # --- END SPECIAL ---


    def _clear_pending() -> None:
        nonlocal pending_dead_boss, pending_dead_event, pending_invalidated
        pending_dead_boss = None
        pending_dead_event = None
        pending_invalidated = False

    def _reset_to_new_pull(begin_ev: Event) -> None:
        nonlocal current_start, current_events, last_combat_end, last_boss_seen_sec
        current_start = begin_ev
        current_events = []
        last_combat_end = None
        last_boss_seen_sec = None
        council_dead.clear()
        _clear_pending()

    def _discard_current_fight() -> None:
        """Jette le fight courant sans le compter."""
        nonlocal current_start, current_events, last_combat_end, last_boss_seen_sec
        current_start = None
        current_events = []
        last_combat_end = None
        last_boss_seen_sec = None
        council_dead.clear()
        _clear_pending()

    def _note_boss_seen(ev: Event) -> None:
        nonlocal last_boss_seen_sec
        if match_boss_name(ev.src) or match_boss_name(ev.dst):
            last_boss_seen_sec = ev.ts_sec

    def _finalize(encounter: str, end_ev: Event) -> None:
        nonlocal kill_idx, current_start, current_events
        if current_start is None:
            return

        kill_idx += 1
        fight = Fight(
            encounter=encounter,
            kill_index=kill_idx,
            start_sec=current_start.ts_sec,
            end_sec=end_ev.ts_sec,
            start_ts=current_start.ts_str,
            end_ts=end_ev.ts_str,
            events=current_events.copy(),
        )
        fight.phases = build_phases_for_fight(fight)
        fights.append(fight)

        _discard_current_fight()

    def _maybe_confirm_pending(now_ev: Event) -> None:
        if current_start is None or not pending_dead_boss or not pending_dead_event:
            return
        if pending_invalidated:
            return
        if now_ev.ts_sec - pending_dead_event.ts_sec >= CONFIRM_DEATH_AFTER_SEC:
            _finalize(pending_dead_boss, pending_dead_event)

    def _trim_recent(now_ts: int) -> None:
        cutoff = now_ts - GLOBAL_BUFFER_SEC
        while recent_events and recent_events[0].ts_sec < cutoff:
            recent_events.popleft()

    def _is_combatlike_boss_interaction(e: Event, boss_key: str) -> bool:
        """
        Détecte une interaction "combat" avant Begin, mais de manière stricte :
        - le boss est impliqué
        - ce n'est pas un event structurel (Begin/End/Kill/Death)
        - raw contient un mot de combat ET un nombre (typique dégâts/soins)
        - l'autre entité n'est pas un boss (évite boss<->boss / scripts)
        """
        if not boss_involved(e, boss_key):
            return False

        if e.code in (0, 11, 12):
            return False
        if getattr(e, "ability", None) in ("Combat Begin", "Combat End"):
            return False

        other = e.dst if match_boss_name(e.src) else e.src
        if match_boss_name(other):
            return False

        raw = (getattr(e, "raw", None) or "")
        if not COMBAT_WORD_RE.search(raw):
            return False
        if not HAS_NUMBER_RE.search(raw):
            return False

        return True

    def _count_boss_interactions_before_begin(boss_key: str) -> int:
        """
        Compte les interactions combat-like dans la fenêtre :
          [begin - LOOKBACK_BEFORE_BEGIN_SEC ; begin)
        """
        if current_start is None:
            return 0

        begin_ts = current_start.ts_sec
        window_start = begin_ts - LOOKBACK_BEFORE_BEGIN_SEC
        cnt = 0

        for e in reversed(recent_events):
            if e.ts_sec >= begin_ts:
                continue
            if e.ts_sec < window_start:
                break
            if _is_combatlike_boss_interaction(e, boss_key):
                cnt += 1
                if cnt >= MIN_COMBATLIKE_INTERACTIONS:
                    return cnt
        return cnt

    def _log_skip(boss_key: str, kill_ev: Event, interactions: int) -> None:
        begin_ts = current_start.ts_str if current_start else "?"
        msg = (
            f"[fight_extractor] Boss '{boss_key}' détecté (kill {kill_ev.ts_str}) mais NON compté: "
            f"Combat Begin trop tardif ({interactions} interaction(s) 'combat-like' "
            f"dans les {LOOKBACK_BEFORE_BEGIN_SEC}s avant Begin {begin_ts})."
        )
        print(msg, file=sys.stderr)

    for ev in events:
        # Buffer global
        recent_events.append(ev)
        _trim_recent(ev.ts_sec)

        _maybe_confirm_pending(ev)

        # Combat Begin
        if ev.code == 0 and ev.ability == "Combat Begin":
            if current_start is None:
                _reset_to_new_pull(ev)
                continue

            # Begin alors qu'un combat est ouvert : rejoin ou nouveau pull
            if last_combat_end is not None:
                gap = ev.ts_sec - last_combat_end.ts_sec
                boss_recent = (
                    last_boss_seen_sec is not None
                    and (ev.ts_sec - last_boss_seen_sec) <= BOSS_RECENT_SEC
                )
                if gap <= REJOIN_GAP_SEC and boss_recent:
                    current_events.append(ev)
                    last_combat_end = None
                    continue

            _reset_to_new_pull(ev)
            continue

        # Hors combat
        if current_start is None:
            continue

        current_events.append(ev)
        _note_boss_seen(ev)

        # Combat End
        if ev.code == 0 and ev.ability == "Combat End":
            last_combat_end = ev
            continue

        # Invalidation pending si le boss réapparait après la "mort"
        if pending_dead_boss and pending_dead_event and ev.ts_sec > pending_dead_event.ts_sec:
            if boss_involved(ev, pending_dead_boss):
                pending_invalidated = True

        # Détection kill boss
        died: Optional[str] = None
        is_kill_line = (ev.code == 11) or bool(KILL_RE.search(ev.raw))
        is_death_line = (ev.code == 12) and bool(DEATH_RE.search(ev.raw))

        if is_kill_line:
            died = ev.dst
        elif is_death_line:
            died = ev.src

        if not died:
            continue

        boss_key = match_boss_name(died)
        if not boss_key:
            continue

        # --- SPECIAL: Council of Fate ---
        # Do NOT finalize on souls/adds, and finalize only when the 3 council members are dead.
        if boss_key == COUNCIL_KEY:
            if not _is_council_member(died):
                # ignore souls / other adds
                continue

            council_dead.add((died or "").strip().casefold())
            if len(council_dead) < 3:
                continue

            # Apply the usual "do not count" rule only on the real end (3rd death).
            if is_kill_line:
                interactions = _count_boss_interactions_before_begin(boss_key)
                if interactions >= MIN_COMBATLIKE_INTERACTIONS:
                    _log_skip(boss_key, ev, interactions)
                    _discard_current_fight()
                    continue

            _finalize(boss_key, ev)
            continue
        # --- END SPECIAL ---

        # Kill => appliquer la règle "ne pas compter" si interactions avant Begin
        if is_kill_line:
            interactions = _count_boss_interactions_before_begin(boss_key)
            if interactions >= MIN_COMBATLIKE_INTERACTIONS:
                _log_skip(boss_key, ev, interactions)
                _discard_current_fight()
                continue

            _finalize(boss_key, ev)
            continue

        # Mort "est mort" => pending confirm
        pending_dead_boss = boss_key
        pending_dead_event = ev
        pending_invalidated = False

    # Fin : confirmer pending si encore valide
    if current_start is not None and pending_dead_boss and pending_dead_event and not pending_invalidated:
        _finalize(pending_dead_boss, pending_dead_event)

    return fights

File: rust/src/fight_extractor.rs
// src/fight_extractor.rs
//
// Détection des kills de boss dans un flux d'événements.
// Port fidèle de fight_extractor.py (PrancingPika) avec ajout de l'approche PT :
//   - Mort détectée via overkill > 0 (langue-neutre, PT) en priorité
//   - Fallback : code 11/12 + regex texte (PP)

use std::collections::{HashSet, VecDeque};

use once_cell::sync::Lazy;
use regex::Regex;

use crate::bosses::{boss_involved, build_phases_for_fight, match_boss_name};
use crate::bosses::council::is_council_member;
use crate::types::{ActionCode, CombatEvent, Fight};

// ─────────────────────────────────────────────────────────────────────────────
// Regex de détection de mort (fallback textuel si overkill == 0)
// Couvre FR et EN — DE également pour robustesse
// ─────────────────────────────────────────────────────────────────────────────
static DEATH_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)(?:est\s+mort[e]?|has\s+died|ist\s+gestorben)").unwrap()
});
static KILL_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)(?:\ba\s+tu[ée]\b|\bhas\s+slain\b|\bhat\s+getötet\b)").unwrap()
});

// Mots typiques d'une ligne de combat (filtre anti-parasite)
static COMBAT_WORD_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)(inflige|dégâts?|touche|frappe|attaque|crit|critique|soigne|damage|heal|hits?|critically|suffers|afflicted)").unwrap()
});
static HAS_NUMBER_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"\b\d+\b").unwrap()
});

// ─────────────────────────────────────────────────────────────────────────────
// Paramètres
// ─────────────────────────────────────────────────────────────────────────────
const REJOIN_GAP_SEC             : u32 = 90;
const BOSS_RECENT_SEC            : u32 = 180;
const CONFIRM_DEATH_AFTER_SEC    : u32 = 6;
const LOOKBACK_BEFORE_BEGIN_SEC  : u32 = 3;
const GLOBAL_BUFFER_SEC          : u32 = 12;
const MIN_COMBAT_INTERACTIONS    : usize = 2;

/// Vérifie si l'événement implique n'importe quel boss (src ou dst)
fn boss_involved_any(ev: &CombatEvent) -> bool {
    match_boss_name(&ev.src).is_some() || match_boss_name(&ev.dst).is_some()
}

/// Mort pending (code 12 "est mort" — à confirmer)
struct PendingDeath {
    boss_key    : String,
    event       : CombatEvent,
    invalidated : bool,
}

// ─────────────────────────────────────────────────────────────────────────────
// État de l'extracteur
// ─────────────────────────────────────────────────────────────────────────────
struct Extractor {
    fights             : Vec<Fight>,
    current_start      : Option<CombatEvent>,
    current_events     : Vec<CombatEvent>,
    kill_idx           : u32,
    last_combat_end    : Option<CombatEvent>,
    last_boss_seen_sec : Option<u32>,
    /// Premier hit impliquant le boss dans ce pull (peut précéder Combat Begin
    /// si le logger est entré en combat après le début effectif du fight)
    first_boss_hit_sec : Option<u32>,
    pending            : Option<PendingDeath>,
    recent             : VecDeque<CombatEvent>,
    council_dead       : HashSet<String>,
}

impl Extractor {
    fn new() -> Self {
        Self {
            fights             : Vec::new(),
            current_start      : None,
            current_events     : Vec::new(),
            kill_idx           : 0,
            last_combat_end    : None,
            last_boss_seen_sec : None,
            first_boss_hit_sec : None,
            pending            : None,
            recent             : VecDeque::new(),
            council_dead       : HashSet::new(),
        }
    }

    // ── Gestion de l'état ─────────────────────────────────────────────────

    fn reset_to_new_pull(&mut self, begin: CombatEvent) {
        // Cherche dans le buffer récent le premier hit boss avant ce Begin
        let begin_ts = begin.ts_sec;
        let earliest = self.recent.iter()
            .filter(|e| e.ts_sec <= begin_ts && boss_involved_any(e))
            .map(|e| e.ts_sec)
            .min();

        self.current_start      = Some(begin);
        self.current_events     = Vec::new();
        self.last_combat_end    = None;
        self.last_boss_seen_sec = None;
        self.first_boss_hit_sec = earliest;
        self.council_dead.clear();
        self.pending            = None;
    }

    fn discard(&mut self) {
        self.current_start      = None;
        self.current_events     = Vec::new();
        self.last_combat_end    = None;
        self.last_boss_seen_sec = None;
        self.first_boss_hit_sec = None;
        self.council_dead.clear();
        self.pending            = None;
    }

    fn note_boss_seen(&mut self, ev: &CombatEvent) {
        if match_boss_name(&ev.src).is_some() || match_boss_name(&ev.dst).is_some() {
            self.last_boss_seen_sec = Some(ev.ts_sec);
        }
    }

    fn note_first_boss_hit(&mut self, ev: &CombatEvent) {
        if boss_involved_any(ev) {
            let t = ev.ts_sec;
            self.first_boss_hit_sec = Some(match self.first_boss_hit_sec {
                Some(prev) => prev.min(t),
                None       => t,
            });
        }
    }

    // ── Finalisation d'un fight ───────────────────────────────────────────

    fn finalize(&mut self, encounter: &str, end_ev: &CombatEvent) {
        let start = match &self.current_start {
            Some(s) => s.clone(),
            None    => return,
        };

        // Utiliser le premier hit boss comme début réel (jamais après Combat Begin)
        let real_start_sec = self.first_boss_hit_sec
            .map(|t| t.min(start.ts_sec))
            .unwrap_or(start.ts_sec);

        self.kill_idx += 1;
        let mut fight = Fight {
            encounter  : encounter.to_string(),
            kill_index : self.kill_idx,
            start_sec  : real_start_sec,
            end_sec    : end_ev.ts_sec,
            start_ts   : start.ts_str.clone(),
            end_ts     : end_ev.ts_str.clone(),
            events     : self.current_events.clone(),
            phases     : Vec::new(),
        };
        fight.phases = build_phases_for_fight(&fight);
        self.fights.push(fight);
        self.discard();
    }

    // ── Confirmation pending ──────────────────────────────────────────────

    fn maybe_confirm_pending(&mut self, now_sec: u32) {
        let should_finalize = match &self.pending {
            Some(p) if !p.invalidated
                && self.current_start.is_some()
                && now_sec.saturating_sub(p.event.ts_sec) >= CONFIRM_DEATH_AFTER_SEC =>
            {
                Some((p.boss_key.clone(), p.event.clone()))
            }
            _ => None,
        };
        if let Some((key, ev)) = should_finalize {
            self.pending = None;
            self.finalize(&key, &ev);
        }
    }

    // ── Buffer circulaire anti-parasite ───────────────────────────────────

    fn trim_recent(&mut self, now_ts: u32) {
        let cutoff = now_ts.saturating_sub(GLOBAL_BUFFER_SEC);
        while self.recent.front().map_or(false, |e| e.ts_sec < cutoff) {
            self.recent.pop_front();
        }
    }

    fn is_combatlike_boss_interaction(&self, ev: &CombatEvent, boss_key: &str) -> bool {
        if !boss_involved(ev, boss_key) { return false; }
        if ev.code_raw == 0 || ev.code == ActionCode::TargetSlain || ev.code == ActionCode::Died {
            return false;
        }
        let other = if match_boss_name(&ev.src).is_some() { &ev.dst } else { &ev.src };
        if match_boss_name(other).is_some() { return false; }
        COMBAT_WORD_RE.is_match(&ev.raw) && HAS_NUMBER_RE.is_match(&ev.raw)
    }

    fn count_boss_interactions_before_begin(&self, boss_key: &str) -> usize {
        let begin_ts = match &self.current_start {
            Some(s) => s.ts_sec,
            None    => return 0,
        };
        let window_start = begin_ts.saturating_sub(LOOKBACK_BEFORE_BEGIN_SEC);
        let mut cnt = 0usize;

        for ev in self.recent.iter().rev() {
            if ev.ts_sec >= begin_ts   { continue; }
            if ev.ts_sec < window_start { break; }
            if self.is_combatlike_boss_interaction(ev, boss_key) {
                cnt += 1;
                if cnt >= MIN_COMBAT_INTERACTIONS { return cnt; }
            }
        }
        cnt
    }

    // ── Détection de qui est mort ─────────────────────────────────────────

    fn detect_death(&self, ev: &CombatEvent) -> Option<(String, bool)> {
        // Priorité 1 : code 11 TargetSlain (langue-neutre)
        if ev.code == ActionCode::TargetSlain {
            return Some((ev.dst.clone(), true)); // true = kill immédiat
        }

        // Priorité 2 : overkill > 0 via spéciaux parsés (PT, langue-neutre)
        // L'attaquant a tué la cible → dst est mort
        if ev.overkill > 0 && ev.is_damage() && !ev.dst.is_empty() {
            return Some((ev.dst.clone(), true));
        }

        // Priorité 3 : KILL_RE sur le texte (fallback FR/EN/DE)
        if KILL_RE.is_match(&ev.raw) {
            return Some((ev.dst.clone(), true));
        }

        // Priorité 4 : code 12 Died (langue-neutre)
        if ev.code == ActionCode::Died {
            return Some((ev.src.clone(), false)); // false = pending
        }

        // Priorité 5 : DEATH_RE sur le texte (fallback)
        if ev.code_raw == 12 || DEATH_RE.is_match(&ev.raw) {
            return Some((ev.src.clone(), false));
        }

        None
    }

    // ── Traitement principal d'un événement ──────────────────────────────

    fn process(&mut self, ev: CombatEvent) {
        // Buffer global
        self.recent.push_back(ev.clone());
        self.trim_recent(ev.ts_sec);

        // Confirmation pending ?
        self.maybe_confirm_pending(ev.ts_sec);

        // ── Combat Begin ───────────────────────────────────────────────
        if ev.is_combat_begin() {
            if self.current_start.is_none() {
                self.reset_to_new_pull(ev);
                return;
            }

            // Begin alors qu'un combat est ouvert → rejoin ou nouveau pull
            if let Some(ref last_end) = self.last_combat_end.clone() {
                let gap = ev.ts_sec.saturating_sub(last_end.ts_sec);
                let boss_recent = self.last_boss_seen_sec
                    .map_or(false, |t| ev.ts_sec.saturating_sub(t) <= BOSS_RECENT_SEC);
                if gap <= REJOIN_GAP_SEC && boss_recent {
                    self.current_events.push(ev);
                    self.last_combat_end = None;
                    return;
                }
            }
            self.reset_to_new_pull(ev);
            return;
        }

        // Hors combat → ignorer
        if self.current_start.is_none() {
            return;
        }

        self.current_events.push(ev.clone());
        self.note_boss_seen(&ev);
        self.note_first_boss_hit(&ev);

        // ── Combat End ────────────────────────────────────────────────
        if ev.is_combat_end() {
            self.last_combat_end = Some(ev);
            return;
        }

        // ── Invalidation pending si le boss réapparait ────────────────
        if let Some(ref mut p) = self.pending {
            if !p.invalidated && ev.ts_sec > p.event.ts_sec {
                if boss_involved(&ev, &p.boss_key.clone()) {
                    p.invalidated = true;
                }
            }
        }

        // ── Détection de mort ────────────────────────────────────────
        let died_info = self.detect_death(&ev);
        let (died_name, is_kill) = match died_info {
            Some(d) => d,
            None    => return,
        };

        let boss_key = match match_boss_name(&died_name) {
            Some(k) => k,
            None    => return,
        };

        // ── SPECIAL : Le Concile du Destin ────────────────────────────
        if boss_key == "Le Concile du Destin" {
            if !is_council_member(&died_name) {
                return; // ignorer âmes et adds
            }
            self.council_dead.insert(died_name.to_lowercase());
            if self.council_dead.len() < 3 { return; }

            if is_kill {
                let n = self.count_boss_interactions_before_begin(boss_key);
                if n >= MIN_COMBAT_INTERACTIONS {
                    eprintln!(
                        "[fight_extractor] Boss '{}' NON compté : Combat Begin trop tardif \
                         ({} interaction(s) dans les {}s avant Begin).",
                        boss_key, n, LOOKBACK_BEFORE_BEGIN_SEC
                    );
                    self.discard();
                    return;
                }
            }
            self.finalize(boss_key, &ev);
            return;
        }

        // ── Kill immédiat (code 11 / overkill / kill_re) ─────────────
        if is_kill {
            let n = self.count_boss_interactions_before_begin(boss_key);
            if n >= MIN_COMBAT_INTERACTIONS {
                eprintln!(
                    "[fight_extractor] Boss '{}' NON compté : Combat Begin trop tardif \
                     ({} interaction(s) dans les {}s avant Begin).",
                    boss_key, n, LOOKBACK_BEFORE_BEGIN_SEC
                );
                self.discard();
                return;
            }
            self.finalize(boss_key, &ev);
            return;
        }

        // ── Mort pending (code 12 / death_re) ────────────────────────
        self.pending = Some(PendingDeath {
            boss_key    : boss_key.to_string(),
            event       : ev,
            invalidated : false,
        });
    }

    // ── Finalisation globale ─────────────────────────────────────────────

    fn finish(mut self) -> Vec<Fight> {
        // Confirmer un pending encore valide en fin de fichier
        if self.current_start.is_some() {
            if let Some(p) = self.pending.take() {
                if !p.invalidated {
                    self.finalize(&p.boss_key.clone(), &p.event.clone());
                }
            }
        }
        self.fights
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// API publique
// ─────────────────────────────────────────────────────────────────────────────
pub fn extract_kills(events: Vec<CombatEvent>) -> Vec<Fight> {
    let mut ext = Extractor::new();
    for ev in events {
        ext.process(ev);
    }
    ext.finish()
}
